packages/dev/core/src/Materials/Node/Blocks/Dual/TBNBlock.ts

```diff
@@ -0,0 +1,171 @@
+    public get TBN(): NodeMaterialConnectionPoint {
+        return this._outputs[0];
+    }
+
+    public get target() {
+        return NodeMaterialBlockTargets.VertexAndFragment;
+    }
+
+    public set target(value: NodeMaterialBlockTargets) {}
+
+    public autoConfigure(material: NodeMaterial) {
+        if (!this.world.isConnected) {
+            let worldInput = material.getInputBlockByPredicate((b) => b.isSystemValue && b.systemValue === NodeMaterialSystemValues.World);
+
+            if (!worldInput) {
+                worldInput = new InputBlock("world");
+                worldInput.setAsSystemValue(NodeMaterialSystemValues.World);
+            }
+            worldInput.output.connectTo(this.world);
+        }
+
+        if (!this.normal.isConnected) {
+            let normalInput = material.getInputBlockByPredicate((b) => b.isAttribute && b.name === "normal");
+
+            if (!normalInput) {
+                normalInput = new InputBlock("normal");
+                normalInput.setAsAttribute("normal");
+            }
+            normalInput.output.connectTo(this.normal);
+        }
+
+        if (!this.tangent.isConnected) {
+            let tangentInput = material.getInputBlockByPredicate((b) => b.isAttribute && b.name === "tangent" && b.type === NodeMaterialBlockConnectionPointTypes.Vector4);
+
+            if (!tangentInput) {
+                tangentInput = new InputBlock("tangent");
+                tangentInput.setAsAttribute("tangent");
+            }
+            tangentInput.output.connectTo(this.tangent);
+        }
+    }
+
+    public prepareDefines(mesh: AbstractMesh, nodeMaterial: NodeMaterial, defines: NodeMaterialDefines) {
+        const normal = this.normal;
+        const tangent = this.tangent;
+
+        let normalAvailable = normal.isConnected && normal.isConnectedToInputBlock;
+        if (normal.connectInputBlock?.isAttribute && !mesh.isVerticesDataPresent(normal.connectInputBlock?.name)) {
+            normalAvailable = false;
+        }
+
+        let tangentAvailable = tangent.isConnected && tangent.isConnectedToInputBlock;
+        if (tangent.connectInputBlock?.isAttribute && !mesh.isVerticesDataPresent(tangent.connectInputBlock?.name)) {
+            tangentAvailable = false;
+        }
+
+        const useTBNBlock = normalAvailable && tangentAvailable;
+
+        defines.setValue("TBNBLOCK", useTBNBlock, true);
+    }
+
+    protected _buildBlock(state: NodeMaterialBuildState) {
+        super._buildBlock(state);
+
+        const normal = this.normal;
+        const tangent = this.tangent;
+        const world = this.world;
+        const TBN = this.TBN;
+
+        if (state.target !== NodeMaterialBlockTargets.Fragment) {
+            // Vertex
+            state._emitVaryingFromString(TBN.associatedVariableName, "mat3");
+
+            state.compilationString += `
+                // ${this.name}
+                vec3 tbnNormal = normalize(${normal.associatedVariableName});

```

Comment:
You should use `${normal.associatedVariableName}.xyz` has we can input a vec4 for the normal.